tactics/eClause.ml

```diff
@@ -176,3 +169,10 @@ let solve_evar_clause env sigma hyp_only clause = function
   in
   let sigma = List.fold_left fold sigma lbind in
   sigma
+
+let check_evar_clause env sigma sigma' eq_clause =
+  let f h = if h.hole_deps then Some (fst (Constr.destEvar (EConstr.Unsafe.to_constr h.hole_evar))) else None in
+  Tacticals.check_evar_list env sigma' (Evar.Set.of_list (List.map_filter f eq_clause.cl_holes)) sigma
+    (fun ?loc evk ->
+       let _, {hole_name} = List.extract_first (fun h -> Evar.equal (fst (Constr.destEvar (EConstr.Unsafe.to_constr h.hole_evar))) evk) eq_clause.cl_holes in
+       Loc.raise ?loc (Logic.RefinerError (env, sigma, UnresolvedBindings [hole_name])))
```

Comment:
This code is horrendous and breaking all the invariants we're ultimately tying to enforce. If you want to have access to the evar itself to perform some checks you could instead expose it in the hole type as a variant of the hole_evar field.

Reply:
Thanks. So, I add a new `hole_evar_key` field, right?

Reply:
Something like that. The current naming scheme is a bit unfortunate, ideally we should have `hole_evar` for the actual evar and rename the current `hole_evar`  field into something something else (`hole_term` maybe?).

Reply:
`hole_term`, yes, why not. (A priori not in the PR though.)